src/typecheck/mod.rs

```diff
@@ -1314,6 +1314,52 @@ fn row_add(
     }
 }
 
+/// Get the set of all unification variables in a type.
+fn get_free_variables(wildcards: &Vec<TypeWrapper>, t: TypeWrapper) -> HashSet<usize> {
+    match t {
+        TypeWrapper::Ptr(i) => HashSet::from([i]),
```

Comment:
I don't think this is sound. If `?i` has been unified with another type with variables inside before, for example `?a -> ?b`, then the free unification variablrs of `?i` shouldn't be `[?i]`, but `[?a,?b]` (if they are fresh, not-yet-unified variables). So, you should follow (and ignore with respect to occur checks, as if the checked variable appears here there is something wrong with the typechecking algorithm) unification variables that point to something in the unification table.

Reply:
I think if `?i` points to `?a -> ?b`, then `?a -> ?b` will be the root of this set because it is no longer a variable: it now a ` TypeWrapper::Concrete(AbsType::Arrow(?a, ?b))`

This is why I call root when checking if two unification variables are the same:
```rust 
fn occurs_check(state: &State, ptr: usize, t: TypeWrapper) -> bool {
    get_free_unif_vars(state.wildcard_vars, t)
        .iter()
        .any(|other_ptr| state.table.root(ptr) == state.table.root(*other_ptr))
}
```

Reply:
It's still not enough, because if `?a` points to `?c -> ?d`, and `?b` is not yet unified, you should have `?c, ?d, ?b` as free variables. Which your `root` call isn't sufficient to catch (it only does one unrolling), if I'm not mistaken.

Even if it would be sufficient, the function is still unsound, because it requires the caller to do something additional which is far from obvious. If it's called `get_free_variables`, so it should return the free unification variables for the standard definition of free unification variables. If someone else comes and refactor the code, something could go wrong. Otherwise, it deserves another name, like impartial, lazy, on_step, etc. and be precisely documented.

Reply:
Yes. I guess this is a another reason to just use a check, instead of generating an intermediate set of free unification variables.

Reply:
I'm not sure I understand why this issue and hashset vs simple check are related. In both cases, I think you will have to potentially perform arbitrary-deep unrolling of free variables?

Reply:
What does "arbitrary-deep unrolling of free variables" mean? Also, see the latest commit to see what I mean by using a simple check instead.